### PR TITLE
Feature: Improved RTP min and max coords + per world settings + claim plugin hooks

### PR DESCRIPTION
**Pull request by DaRacci**
This PR allows RTP min and max settings to be defined per world as well as using what I believe is a much nicer looking approach to the config settings of it.

Note this has not been tested yet and will be soon but I will update this PR once it has been tested.

### CODE REVIEW

**Comment by Thatsmusic99** [comment]
Although a draft, this does look very promising. I've always wanted to get a better approach to the RTP coordinates.

Best of luck with it and let me know if you have any questions regarding conventions or any API within the plugin!

**Comment by DaRacci** [comment]
> Although a draft, this does look very promising. I've always wanted to get a better approach to the RTP coordinates.
> 
> Best of luck with it and let me know if you have any questions regarding conventions or any API within the plugin!

Yeah I was just configuring the plugin for my own server and noticed the lack of per world options besides for border plugin integration so I thought id see if I could wip something up.

**Comment by DaRacci** [comment]
Actually I have just noticed I made it use the X coords if Z isnt present but should this check for default before reusing those do you think?

**Comment by Thatsmusic99** [comment]
I assume when you say that, you mean that when the Z coordinates for a world aren't present, the X coordinate for the world is used instead?

Huh. That's actually quite a good question - I think it would be worth using the present coordinates in X (or even Z) to create a square radius, but even then, it may be worth documenting that in the config file to avoid confusion.

**Comment by DaRacci** [comment]
> I assume when you say that, you mean that when the Z coordinates for a world aren't present, the X coordinate for the world is used instead?
> 
> Huh. That's actually quite a good question - I think it would be worth using the present coordinates in X (or even Z) to create a square radius, but even then, it may be worth documenting that in the config file to avoid confusion.

I'll add a note about it in the config and I mean if it does need to be changed its a fairly easy thing to do in the end

**Comment by DaRacci** [comment]
I saw that claim support was on the project board so i thought id give that a crack too, still untested but its late so im gonna leave that to tomorrow

**Comment by DaRacci** [comment]
With the introduction of claim protection in rtp would it be better to a) ditch checking if that player is trusted there at all, b) prepare the area in rapid response without checking claims and then check if they are trusted during the command, c) Prepare the rapid response avoiding claims and then without rapid response check if the player is trusted since the location is found at the same time.

**Comment by Thatsmusic99** [comment]
I assume it would be best to check if there's any claim and avoid it altogether, even if you're trusted in that claim, you don't want to necessarily randomly TP into your friend's house because it's meant to be in a different location.

**Comment by DaRacci** [comment]
Thats a good point, ill implement it that way

**Comment by Thatsmusic99** [comment]
If it's okay as well, may I ask if you're okay with me switching the base branch for this to v5.7? Figured I'd want one place to have all the branches merged into so I can test it before pushing it into the master branch, just want to confirm that's fine with you though and that it won't cause the end of the world.

**Comment by DaRacci** [comment]
yeah that's all good

**Comment by Thatsmusic99** [comment]
Cheers, will get that done now!

**Comment by DaRacci** [comment]
Where abouts is the location gotten when not using rapid response i'm having some issues actually tracking it down to add the claim check

**Comment by DaRacci** [comment]
Okay so, I've now tested this with the lands plugin and it's fully functional and working i see if i can figure out the other two but i've never actually used them myself

**Comment by DaRacci** [comment]
while testing the different values per world i've noticed that it seems like the  rapid response locations disregard the world, as when i first testing in the nether i was getting up to 5000 blocks of travel but after 3 attempts it went to 500 like i set it.

**Comment by Thatsmusic99** [comment]
Checking this, the `locQueue` in RTPManager may need resetting between plugin reloads (not necessarily server restarts but I may reconsider), it currently keeps the cached locations even after the plugin is reloaded and that shouldn't be intended.

**Comment by DaRacci** [comment]
Alright i can confirm all three claim hooks are working, which should conclude this PR i think.

**Comment by Thatsmusic99** [comment]
Perfect! My work shift does start in 5 minutes so I'll review this once I'm off the clock.

**Comment by DaRacci** [comment]
Sounds good, Hopefully i tested it all properly lol

**Review by Thatsmusic99 — COMMENTED** [review]
From first impressions, this looks perfect! Will be testing it shortly to ensure everything does indeed work.

File: AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/managers/PluginHookManager.java [hunk 1]
```diff
@@ -45,34 +59,27 @@ public ImportExportPlugin getImportPlugin(String name) {
...
 
-    private void loadImportPlugin(String name, Class<? extends ImportExportPlugin> clazz) {
+    private <T> void loadPlugin(HashMap<String, T> map, String name, Class<? extends T> clazz) {
         try {
-            importPlugins.put(name, clazz.newInstance());
-        } catch (InstantiationException | IllegalAccessException e) {
+            map.put(name, clazz.getConstructor().newInstance());
+        } catch (InstantiationException | NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
             e.printStackTrace();
-        } catch (NoClassDefFoundError ignored) {
```
Comment (Thatsmusic99) [inline AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/managers/PluginHookManager.java:53]: As it turns out, removing the check for this exception does break the plugin, e.g. if Essentials is not installed:

```java.lang.NoClassDefFoundError: com/earth2me/essentials/commands/WarpNotFoundException```

Ironically, this is the only plugin that throws such a fit.

**Review by Thatsmusic99 — CHANGES_REQUESTED** [review]
One very small bug that should be easy to fix, but otherwise, this works perfectly for me!

**Comment by DaRacci** [comment]
With that fixed it should be completed now then

**Review by Thatsmusic99 — APPROVED** [review]
Works fine now, appreciate the commit energy.

Thank you for contributing to AdvancedTeleport!

**Pull request merged**